Feature: Config Remote API

dungltcn272 opened the pull request:
## Related Tickets
- [Feature: Config Remote API] (https://edu-redmine.sun-asterisk.vn/issues/90694)
## WHAT
- Set ApiEndPoint
- Identìy Dto (Response)
- Code parse function to get model from json
- Code MovieRemoteDataSource using to get data from api (not model used for UI)

## Evidence (Screenshot or Video)
<img width="397" height="705" alt="image" src="https://github.com/user-attachments/assets/baed7050-e28b-45a0-ab16-c2dcaedcb666" />


## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

daolq-2712 committed:
Create README.md

daolq-2712 committed:
MVP directory structure

dungltcn272 committed:
Merge branch 'awesome-academy:develop' into develop

dungltcn272 committed:
Add BaseVIew, BasePresenter

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/MovieRepository.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/MovieRepositoryImpl.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Merge branch 'awesome-academy:develop' into develop

dungltcn272 committed:
Merge remote-tracking branch 'origin/develop' into feat_config_remote

# Conflicts:
#	app/src/main/java/com/sun/moviedb/screen/MainActivity.kt
#	app/src/main/java/com/sun/moviedb/screen/detail/MovieDetailFragment.kt
#	app/src/main/java/com/sun/moviedb/utils/navigation/AppNavigator.kt
#	app/src/main/java/com/sun/moviedb/utils/navigation/NavDestination.kt
#	app/src/main/res/layout/fragment_home.xml
#	app/src/main/res/layout/fragment_movie_detail.xml

dungltcn272 committed:
ApiEndpoint

dungltcn272 committed:
Create dto model suitable for api

dungltcn272 committed:
Create Parse Json to dto models function

dungltcn272 committed:
Avoid NullPointerException (assign null if key not found in jsonObject)

dungltcn272 committed:
MovieRemoteDataSource.kt for get Dto data from api (not model for UI)

noname1288 commented:
Bác ơi, e nghĩ: để tránh NullExceptionPointer thì nên khởi tạo default value trong data class (trường hợp nào api trả về null thì set null)

dungltcn272 committed:
Merge remote-tracking branch 'origin/feat_config_remote' into feat_config_remote

dungltcn272 committed:
Avoid NullPointerException (assign null if key not found in jsonObject)

noname1288 committed:
update model, dto

noname1288 committed:
- set default in models
- update Api Helper: not handle threads in here (hanle in MovieRemoteDataSource)
- update MovieRemoteDataSource
- update MovieRepository.kt

noname1288 committed:
Update MovieDataSource.kt

noname1288 committed:
Update Endpoint.kt

dungltcn272 committed:
[Quang] Improve Config with set default in model

dungltcn272 committed:
Set all api function

dungltcn272 committed:
Set all api function

dungltcn272 committed:
Merge remote-tracking branch 'origin/feat_config_remote' into feat_config_remote

# Conflicts:
#	app/src/main/java/com/sun/moviedb/data/repository/source/MovieDataSource.kt
#	app/src/main/java/com/sun/moviedb/data/repository/source/MovieRepository.kt
#	app/src/main/java/com/sun/moviedb/data/repository/source/remote/MovieRemoteDataSource.kt
#	app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt

dungltcn272 committed:
Clean some class package

daolq-2712 commented on app/build.gradle.kts:
```diff
@@ -40,6 +40,7 @@ android {
 }
 
 dependencies {
+    testImplementation(libs.mockwebserver)
```
```suggestion
```
pj 1 này mình chưa làm test

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/model/Movie.kt:
```diff
@@ -1,3 +1,30 @@
+//    val sub_docquyen: Boolean = false,
+    val thumb_url: String = "",
+    val time: String = "",
+//    val trailer_url: String = "",
```
sao lại comment line code này lại e nhỉ? mình có chức năng xem video trail mà?

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/repository/source/remote/NetworkResult.kt:
```diff
@@ -0,0 +1,9 @@
+        val code: Int?,
+        val message: String
+    ) : NetworkResult<Nothing>()
+}
```
```suggestion
}

```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/model/Movie.kt:
```diff
@@ -1,3 +1,30 @@
 
-data class Movie(val id : Int)
+data class Movie(
+    val _id: String = "",
```
```suggestion
    val id: String = "",
```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/model/Data.kt:
```diff
@@ -0,0 +1,11 @@
+    val items: List<Item> = emptyList(),
+    val params: Params = Params(),
+    val titlePage: String = "",
+    val type_list: String = ""
```
```suggestion
    val typeList: String = ""
```

daolq-2712 replied:
đặt tên biến trong PJ nên thống nhất dùng 1 kiểu e nhé,
Kotlin thường sẽ dùng **camel case** sửa tương tự ở những model khác e nhé.

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/model/Episode.kt:
```diff
@@ -0,0 +1,6 @@
+
+data class Episode(
+    val server_data: List<ServerData> = emptyList(),
+    val server_name: String = ""
```
```suggestion
    val serverData: List<ServerData> = emptyList(),
    val serverName: String = ""
```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/model/Movie.kt:
```diff
@@ -1,3 +1,30 @@
+    val _id: String = "",
+    val actor: List<String> = emptyList(),
+    val category: List<Category> = emptyList(),
+    val chieurap: Boolean = false,
```
gửi sample json response của API trả về, anh việt lộn xộn như này anh thấy sai sai?

noname1288 replied:
{
    "status": true,
    "msg": "done",
    "movie": {
        "tmdb": {
            "type": "movie",
            "id": "1234821",
            "season": null,
            "vote_average": 6.365,
            "vote_count": 640
        },
        "imdb": {
            "id": null
        },
        "created": {
            "time": "2025-07-17T06:11:37.000Z"
        },
        "modified": {
            "time": "2025-08-05T22:12:48.000Z"
        },
        "_id": "2da535ad78bb2e93aa448b1a4a61134e",
        "name": "Thế Giới Khủng Long: Tái Sinh",
        "slug": "the-gioi-khung-long-tai-sinh",
        "origin_name": "Jurassic World: Rebirth",
        "content": "Phim lấy bối cảnh 5 năm sau phần phim Thế Giới Khủng Long: Lãnh Địa (Jurassic World: Dominion). Thế Giới Khủng Long: Tái Sinh mở ra một chương mới đầy tính hành động, chứng kiến một đội khai thác chạy đua đến nơi nguy hiểm nhất trên Trái Đất. Dàn nhân vật chính là bộ ba Scarlett Johansson, Mahershala Ali và Jonathan Bailey dấn thân vào một nhiệm vụ cực kỳ hiểm nguy, đó chính là cố gắng lấy DNA có thể dẫn đến một bước đột phá y học cho nhân loại. Chìa khóa của nó tình cờ lại là DNA của ba con khủng long khổng lồ nhất trên cạn, biển và không trung trong sinh quyển nhiệt đới. Hành trình này sẽ đưa nhóm nhân vật chính băng rừng, vượt biển, đối mặt với nhiều loài khủng long kỳ lạ, nguy hiểm nhưng cũng đầy lý thú, từ đó hé mở nhiều điều bí ẩn mà tạo hóa đã giấu khỏi con người suốt bấy lâu nay.",
        "type": "single",
        "status": "completed",
        "poster_url": "https://phimimg.com/upload/vod/20250717-1/f0b3a51863368ae2706f0844c9bf3226.jpg",
        "thumb_url": "https://phimimg.com/upload/vod/20250717-1/8cbdfb269ca3805d0f2705f7ea22a2c5.jpg",
        "is_copyright": false,
        "sub_docquyen": false,
        "chieurap": true,
        "trailer_url": "https://www.youtube.com/watch?v=2ZhB-YO5Tnk",
        "time": "134 phút",
        "episode_current": "Full",
        "episode_total": "1",
        "quality": "FHD",
        "lang": "Vietsub",
        "notify": "",
        "showtimes": "",
        "year": 2025,
        "view": 0,
        "actor": [
            "Scarlett Johansson",
            "Jonathan Bailey",
            "Mahershala Ali",
            "Rupert Friend",
            "Manuel Garcia-Rulfo",
            "Luna Blaise",
            "David Iacono",
            "Audrina Miranda",
            "Bechir Sylvain",
            "Philippine Velge",
            "Ed Skrein",
            "Adam Loxley",
            "Niamh Finlay",
            "Julian Edgar",
            "Lucy Thackeray",
            "Billy Smith",
            "Jonny L"
        ],
        "director": [
            "Gareth Edwards"
        ],
        "category": [
            {
                "id": "0bcf4077916678de9b48c89221fcf8ae",
                "name": "Khoa Học",
                "slug": "khoa-hoc"
            },
            {
                "id": "68564911f00849030f9c9c144ea1b931",
                "name": "Viễn Tưởng",
                "slug": "vien-tuong"
            },
            {
                "id": "66c78b23908113d478d8d85390a244b4",
                "name": "Phiêu Lưu",
                "slug": "phieu-luu"
            },
            {
                "id": "9822be111d2ccc29c7172c78b8af8ff5",
                "name": "Hành Động",
                "slug": "hanh-dong"
            },
            {
                "id": "a7b065b92ad356387ef2e075dee66529",
                "name": "Tâm Lý",
                "slug": "tam-ly"
            }
        ],
        "country": [
            {
                "id": "74d9fa92f4dea9ecea8fc2233dc7921a",
                "name": "Âu Mỹ",
                "slug": "au-my"
            }
        ]
    },
    "episodes": [
        {
            "server_name": "#Hà Nội (Vietsub)",
            "server_data": [
                {
                    "name": "Full",
                    "slug": "full",
                    "filename": "Thế Giới Khủng Long: Tái Sinh - Jurassic World: Rebirth - 2025 - 1080p - Vietsub - Full",
                    "link_embed": "https://player.phimapi.com/player/?url=https://s6.kkphimplayer6.com/20250805/A6Gw4nda/index.m3u8",
                    "link_m3u8": "https://s6.kkphimplayer6.com/20250805/A6Gw4nda/index.m3u8"
                }
            ]
        }
    ]
}

noname1288 replied:
e gửi anh phần json response. Lý do lộn xộn là trong quá trình làm, e sử dụng plugin (Json to Kotlin class: tự động gen file json -> data class) a ạ

daolq-2712 commented on app/src/main/java/com/sun/moviedb/screen/home/HomeFragment.kt:
```diff
@@ -2,6 +2,7 @@ package com.sun.moviedb.screen.home
 
 import android.view.LayoutInflater
 import android.view.ViewGroup
+import com.sun.moviedb.data.repository.source.remote.MovieRemoteDataSource
```
```suggestion
```
un used import

daolq-2712 commented on app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt:
```diff
@@ -1,109 +1,102 @@
+                val list = parser(body)
+                NetworkResult.OnSuccess(list)
+            } else {
+                NetworkResult.OnError(code, "HTTP $code: $body")
```
@dungltcn272 
anh đọc code thấy 2 hàm [getListFromUrl](https://github.com/awesome-academy/HN_Pra_2025_Android_MovieDB/pull/4/files#diff-daa02062f8c146a9d508d5df5bbc66d6c561f31bca762b7f96a703f0907f4250R78) & [getObjectFromUrl](https://github.com/awesome-academy/HN_Pra_2025_Android_MovieDB/pull/4/files#diff-daa02062f8c146a9d508d5df5bbc66d6c561f31bca762b7f96a703f0907f4250R52) giống nhau hoàn toàn sao phải đặt tên khác nhau nhỉ?

daolq-2712 commented on app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt:
```diff
@@ -1,109 +1,102 @@
-    }
+    data class RawData(val code: Int, val body: String?)
 
-    private val executor = Executors.newSingleThreadExecutor()
-    private val mainHandler = Handler(Looper.getMainLooper())
```
lý do tại sao em xoá những dòng xử lý async này ở đây & move ra ngoài e nhỉ?
Xử lý code base API mình nên giữ ở đây

noname1288 replied:
Bởi vì e nghĩ tại RemoteSource có thể gọi 2 api song song & tổng hợp kết quả. Nên e mới đưa ra ngoài a ạ

daolq-2712 replied:
Không liên quan đến xử lý gọi 2 API song song & tổng hợp kết quả, ca này em dùng AI Agent để suggest code đúng không :))

daolq-2712 commented on app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt:
```diff
@@ -1,109 +1,102 @@
-    private val executor = Executors.newSingleThreadExecutor()
-    private val mainHandler = Handler(Looper.getMainLooper())
+    @Throws(Exception::class)
+    fun getRawDataFromURL(
```
```suggestion
    private fun getRawDataFromURL(
```

daolq-2712 reviewed (changes requested):
Pls check my comments!

dungltcn272 committed:
Fix
Camel case
Replace getListFromUrl & getObjectFromUrl by GetResultFromUrl in ApiHelper.kt
Async Executors.newFixedThreadPool(4) in ApiHelper.kt

dungltcn272 commented:
All well done

daolq-2712 commented on app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt:
```diff
@@ -2,108 +2,62 @@ package com.sun.moviedb.utils.network
+object ApiHelper {
+    private const val TIME_OUT = 15000
+    private const val METHOD_GET = "GET"
+    // asynchronous executor 4 threads
+    private val executor = Executors.newFixedThreadPool(4)
```
```suggestion
    private val executor = Executors.newSingleThreadExecutor()
```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/repository/source/remote/parse/ParseCountryList.kt:
```diff
@@ -0,0 +1,10 @@
+    return (0 until arr.length()).map { i ->
+        arr.getJSONObject(i).toCountry()
+    }
+}
```
```suggestion
}

```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/repository/source/remote/dto/MovieListResponse.kt:
```diff
@@ -0,0 +1,12 @@
+    val pagination: Pagination,
+    val status: Boolean
+)
+
```
```suggestion
```

daolq-2712 commented on app/src/main/java/com/sun/moviedb/data/repository/source/MovieRepository.kt:
```diff
@@ -0,0 +1,93 @@
+            }
+    }
+}
+
```
```suggestion
```

daolq-2712 reviewed (commented):
Pls check my comments!

dungltcn272 commented:
E có học hỏi 1 chút thôi ạ :((

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/source/remote/parse/ParseCountryList.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/utils/network/ApiHelper.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/source/remote/NetworkResult.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/source/remote/dto/MovieListResponse.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>

dungltcn272 committed:
Update app/src/main/java/com/sun/moviedb/data/repository/source/MovieRepository.kt

Co-authored-by: Le Quang Dao <90740771+daolq-2712@users.noreply.github.com>
